Added new Domains and Updated code with comments

topleansebastian opened the pull request.

topleansebastian committed:
Added new Operations and added new comments

Updated Program.cs
Added 6 new Domains
Added CreateClient Domain, not implemented yet
Updated Hardcoded file
Sessions will be tracked via Dictionary<String: SessionID, Cart: Cart>

topleansebastian committed:
Added comments in Program.cs

topleansebastian committed:
Added most of the Employee domains and updated Program.cs

Added new Domains for the employee
- Can set a menu to be vizible during certain hours;
- Can get all the current orders from a restaurant;
- Can set an order status of a cart;
- Can request payment from a client;
- Can check the payment status of a client;

Updated Program.cs to include the new domains

cosmintiru commented on Toplean_Alexandru/week-02/Domain/Domain/CreateMenuOp/CreateMenuOp.cs:
```diff
@@ -23,8 +23,10 @@ public override Task<ICreateMenuResult> Work(CreateMenuCmd Op, Unit state)
                 if (CommandIsValid)
                 {
-                    Menu menu = new Menu(Op.Name, Op.MenuType);
+                    Menu menu = new Menu(Op.Name, Op.MenuType, Op.MenuVisibilityTypes, Op.Hour);
                     Op.Restaurant.Menus.Add(menu);
+                    if (Op.MenuVisibilityTypes == MenuVisibilityTypes.SpecialMenu)
+                        AllHardcodedValues.HarcodedVals.SpecialMenus[Op.Hour].Add(menu);
```
Since you've used this static class to emulate persistence, please make sure the operation is idempotent. What should happen if I call the CreateMenuOp twice, with the same arguments? What should be the outcome of that?

cosmintiru commented on Toplean_Alexandru/week-02/Domain/Domain/GetOrdersOp/GetOrdersCmd.cs:
```diff
@@ -0,0 +1,31 @@
+        public GetOrdersCmd(Restaurant restaurant)
+        {
+            Restaurant = restaurant;
+            Orders = restaurant.Orders;
```
Probably there's a bit of confusion here, the command itself should not contain the actual result of the operation. Yes, it takes a restaurant as an input, but it's the Ops responsibility to resolve the orders and return them (not command's responsibility).

cosmintiru commented on Toplean_Alexandru/week-02/Domain/Domain/DeleteRestaurantOp/DeleteRestaurantOp.cs:
```diff
@@ -16,12 +16,17 @@ public override Task<IDeleteRestaurantResult> Work(DeleteRestaurantCmd Op, Unit
             if (Exists(Op.RestaurantName))
             {
+                // Get the restaurant that will be deleted
                 Restaurant restaurant = AllHardcodedValues.HarcodedVals.RestaurantList.FirstOrDefault(r => r.Name == Op.RestaurantName);
+                // Deleting the restaurant
                 AllHardcodedValues.HarcodedVals.RestaurantList.Remove(restaurant);
```
What happens if the restaurant doesn't exist? FirstOrDefault will return null, which will most likely blow up your code. Should you adapt your response to this case?

cosmintiru commented on Toplean_Alexandru/week-02/Domain/Domain/GetMenusOp/GetMenusCmd.cs:
```diff
@@ -0,0 +1,22 @@
+
+        public GetMenusCmd(Restaurant restaurant)
+        {
+            Restaurant = restaurant;
+            Menus = restaurant.Menus;
+        }
```
Same as for GetOrdersCmd.

cosmintiru commented on Toplean_Alexandru/week-02/Domain/Domain/GetMenusOp/GetMenusOp.cs:
```diff
@@ -0,0 +1,38 @@
+        {
+            if (Exists(Op.Restaurant))
+            {
+                if (Op.IsValid())
+                {
+                    // Menus are successfully selected
+                    return Task.FromResult<IGetMenusResult>(new MenusGot(Op.Menus));
+                }
+                else
+                {
+                    // No menus were selected, Menu List of the restaurant was null
+                    return Task.FromResult<IGetMenusResult>(new NoMenusGot($"Restaurant {Op.Restaurant.Name} has no menus"));
```
This is where you should implemented the retrieval of Menus. This is where the logic should reside, instead of the GetMenusCmd constructor.

cosmintiru commented on Toplean_Alexandru/week-02/Domain/Domain/PlaceOrderOp/PlaceOrderOp.cs:
```diff
@@ -0,0 +1,25 @@
+            (bool CommandIsValid, String Error) = Op.IsValid();
+            if (CommandIsValid)
+            {
+                AllHardcodedValues.HarcodedVals.Carts[Op.Client.SessionID].Status = Models.Cart.CartStatus.CartSubmitted;
```
What happens if your persistence layer doesn't have that cart? 

https://twitter.com/billwagner/status/1251156770988232705

cosmintiru commented on Toplean_Alexandru/week-02/Domain/Domain/RequestPaymentOp/RequestPaymentOp.cs:
```diff
@@ -0,0 +1,25 @@
+            if (Op.IsValid())
+            {
+                // Sets the cart status to PaymentRequested to the client with the session ID = SessionID;
+                AllHardcodedValues.HarcodedVals.Carts[Op.SessionID].Payment = Models.Cart.PaymentStatus.PaymentRequested;
```
Unsafe indexing?

cosmintiru reviewed (commented):
When I tried to compile your code most of the new "result" types lacked "using System.Threading.Tasks" which makes me believe that your code doesn't run locally. However, fixing those and running your expression, to some extent it did what it promised. 

There are a couple of overall mentions:

1. Your validation could probably be simplified. However, the way you implement them it's up to you. Can you try the annotation option too? If you annotate your models/commands your validations will have no cyclomatic complexity in most of the cases.

2. There are a couple of Commands that contain logic. It may not seem so, but attaching the Orders, for example, in the command constructor and use that field in the operation, counts as business logic. 

3. It's great that you used that hardcoded store to emulate a persistence layer. However, please make sure that it doesn't inflict unexpected side effects when you run your code in unexpected ways. (i.e. you have a transient fault, the http request doesn't properly complete, the request is issued twice, your expression runs twice).

topleansebastian committed:
Added CreateCartItem Domain, updated bad code

- Added CreateCartItem Domain
- Trying to use mainly CreateCartItem Domain rather than CreateOrder Domain
- Added a few attributes
- Added some new bugs to discover later
- Added using System.Threading.Tasks to all "results"
- Fixed bad code
- All IsValid() functions from all commands will return a (bool, String) tuple
- Updated Program.cs to add a new CartItem aswell

topleansebastian committed:
Started implementing a menu in Program.cs and added CreateClient Domain

topleansebastian committed:
Updated Program.cs menus and Added new Select Domains

- Updated Program.cs
- Added SelectMenu Domain
- Added GetMenuItem Domain
- Small changes in Select Client to be selected only by SessionID and not by Username

topleansebastian committed:
Added Get Cart Item Domain and updated Program.cs Menu

- Implemented some options for the employee menu
- Implemented most of the client menu options
- Added GetCartItem Domain
- Trying to modify the Order class to have a Cart member rather than a list of MenuItems
